### ISSUE TITLE
`unused_unit` reports confusing suggestion for trait objects with unnecessary parenthesisin the type

### ISSUE THREAD

**Issue by sdroege** [rust-lang/rust-clippy#15176]
### Summary

See https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=6e4227dddcac03952423666db31f2780

### Lint Name

unused_unit

### Reproducer

I tried this code:

```rust
fn main() {
    let _x: Box<&dyn (FnMut(i32))> = todo!();
}
```

I saw this happen:

```
warning: unneeded unit return type
 --> src/main.rs:2:23
  |
2 |     let _x: Box<&dyn (FnMut(i32))> = todo!();
  |                       ^^^^^^^^^^ help: remove the `-> ()`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unused_unit
  = note: `#[warn(clippy::unused_unit)]` on by default
```

I expected to see this happen: A suggestion to remove the unnecessary parenthesis

### Version

```text
rustc 1.88.0 (6b00bc388 2025-06-23)
binary: rustc
commit-hash: 6b00bc3880198600130e1cf62b8f8a93494488cc
commit-date: 2025-06-23
host: x86_64-unknown-linux-gnu
release: 1.88.0
LLVM version: 20.1.5
```

**Comment by samueltardieu** [comment]
Thanks for your report. This has been fixed by #14962, and the fix is in the beta version already (rust 1.89.0-beta.2).

**Comment by samueltardieu** [comment]
(note: there was no corresponding bug report, so it is expected that you would not find it by looking at issues)

**Comment by sdroege** [comment]
Thanks!
